geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,210 @@
+      memberBuckets.computeIfAbsent(memberId, k -> new ArrayList<>()).add(bucketId);
+    }
+
+    return memberBuckets;
+  }
+
+  /**
+   * This returns a list of {@link MemberBucketSlot}s where each entry corresponds to a bucket. If
+   * the details for a given bucket cannot be determined, that entry will contain {@code null}.
+   */
+  public synchronized List<MemberBucketSlot> getBucketSlots() {
+    initializeBucketsIfNecessary();
+
+    List<MemberBucketSlot> memberBucketSlots = new ArrayList<>(RegionProvider.REDIS_REGION_BUCKETS);
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      Pair<String, Integer> hostPort = getHostPort(bucketId);
+      if (hostPort != null) {
+        memberBucketSlots.add(
+            new MemberBucketSlot(bucketId, hostPort.getLeft(), hostPort.getRight()));
+      }
+    }
+
+    return memberBucketSlots;
+  }
+
+  private InternalDistributedMember getOrCreateMember(int bucketId) {
+    return dataRegion.getOrCreateNodeForBucketWrite(bucketId, null);
+  }
+
+  private void initializeBucketsIfNecessary() {
+    if (dataRegion.getDataStore() != null &&
+        dataRegion.getDataStore().getAllLocalBucketIds().isEmpty()) {
+      PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+    }
+  }
+
+  /**
+   * This method will retry for {@link #HOSTPORT_RETRIEVAL_ATTEMPTS} attempts and return null if
+   * no information could be retrieved.
+   */
+  private Pair<String, Integer> getHostPort(int bucketId) {
+    Pair<String, Integer> response;
+
+    for (int i = 0; i < HOSTPORT_RETRIEVAL_ATTEMPTS; i++) {
+      response = getHostPort0(bucketId);
+      if (response != null) {
+        return response;
+      }
+
+      try {
+        Thread.sleep(HOSTPORT_RETRIEVAL_INTERVAL);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    logger.error("Unable to retrieve host and redis port for member with bucketId: {}", bucketId);
+
+    return null;
+  }
+
+  @SuppressWarnings("unchecked")
+  private Pair<String, Integer> getHostPort0(int bucketId) {
+    InternalDistributedMember member = getOrCreateMember(bucketId);
+
+    if (hostPorts.containsKey(member)) {
+      return hostPorts.get(member);
+    }
+
+    List<RedisMemberInfo> memberInfos;
+    try {
+      ResultCollector<RedisMemberInfo, List<RedisMemberInfo>> resultCollector =
+          FunctionService.onRegion(dataRegion).execute(RedisMemberInfoRetrievalFunction.ID);
+      memberInfos = resultCollector.getResult();
+    } catch (Exception e) {
+      logger.warn("Unable to execute {}: {}", RedisMemberInfoRetrievalFunction.ID, e.getMessage());
+      return null;
+    }
+
+    hostPorts.clear();
+
+    for (RedisMemberInfo memberInfo : memberInfos) {
+      Pair<String, Integer> hostPort =
+          Pair.of(memberInfo.getHostAddress(), memberInfo.getRedisPort());
+      hostPorts.put(memberInfo.getMember(), hostPort);
```

Comment:
Huh, should hostPorts just be a map of <String, RedisMemberInfo> ? Might be a bit clearer than Pair<Integer, String>

Reply:
Changed.